audio-worklet/noisegate-test.html

```diff
@@ -0,0 +1,432 @@
+-->
+<!DOCTYPE html>
+  <head>
+    <script src="test/testharness.js"></script>
+    <script src="test/testharnessreport.js"></script>
+    <script src="test/audit-util.js"></script>
+    <script src="test/audit.js"></script>
+    <script src="src/noisegate-script-processor.js"></script>
+
+    <script id="noisegate-test-code">
+      // Sample rate is a power of two to avoid interpolation from delay node.
+      // It is held artificially low for easy inspection of audio buffers.
+      const sampleRate = 4096;
+      const testDurationSec = 1;
+      const testFrames = testDurationSec * sampleRate;
+      const bufferSize = 512;
+      const scriptProcessorLatency = bufferSize * 2;
+      let audit = Audit.createTaskRunner();
+      
+      // With a threshold of t and a constant source of some value above t,
+      // followed by a sequence of values beneath t, an ideal envelope
+      // follower would make the noise gate output equal to the input. But
+      // a proper envelope can only be approximated. This test verifies
+      // that the envelope delay is mathematically correct given alpha.
+      audit.define('envelope-is-mathematically-correct', (task, should) => {
+        let context = new OfflineAudioContext(2, testFrames, sampleRate);
+        const timeToThreshold = 0.25;
+
+        // The threshold is scaled up by the square root of two since the
+        // envelope is tuned to sine waves.
+        const linearThreshold = 0.5 * Math.sqrt(2);
+
+        // The noise gate bases muting decisions on a decibel scale.
+        const threshold = toDecibel_(linearThreshold);
+
+        // Attack and release are both 0.
+        let noisegate = new NoiseGate(context, {bufferSize, threshold});
+
+        // The signal should be all 1's prior to timeToThreshold to simplify
+        // the math that relates the closing index to alpha.
+        let gain = new GainNode(context, {gain : 1});
+        let osc = new ConstantSourceNode(context, {offset: 1});
+        osc.connect(gain).connect(noisegate.input);
+        noisegate.output.connect(context.destination);
+
+        // The post-threshold gain should make samples less than the threshold.
+        // It must be greater than 0; otherwise samples will immediately
+        // be 0 after the threshold, and there will be no inspectable delay.
+        const beneathThresholdGain = 0.1;
+        gain.gain.setValueAtTime(0.1, timeToThreshold);
+        
+        osc.start();
+
+        context.startRendering().then((buffer) => {
+          let channel = buffer.getChannelData(0);
+          
+          // With a constant source, the number of steps it takes for the
+          // envelope to open and close is a function of alpha relative
+          // to the threshold.
+          const numOpeningSteps =
+              getNumStepsFromZeroUpTo_(linearThreshold, noisegate.alpha_);
+
+          // If sample rate or alpha changes, add one to numClosingSteps change.
+          const numClosingSteps =
```

Comment:
Some very brief comments here on what you're trying to test and how would be really helpful.